### PR TITLE
Support passing data in numpy int8, int16, uint8 and uint16 dtypes to GMT

### LINKED ISSUE #1960: uint8 datatype issue
**Description of the problem**

 I read an image of uint8 datatype with `rioxarray.open_rasterio` and plot with pygmt raise errors.

If I convert datatype to float, my script runs fluently.

**Full error message**

```
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py", line 3552, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-cbda221bbf05>", line 1, in <module>
    runfile('D:/NutSync/code/funcs_lib_dg/plot_func/plot_raster_pygmt.py', wdir='D:/NutSync/code/funcs_lib_dg/plot_func')
  File "C:\Program Files\JetBrains\PyCharm 2020.1\plugins\python\helpers\pydev\_pydev_bundle\pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm 2020.1\plugins\python\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "D:/NutSync/code/funcs_lib_dg/plot_func/plot_raster_pygmt.py", line 307, in <module>
    xrd = main()
  File "D:/NutSync/code/funcs_lib_dg/plot_func/plot_raster_pygmt.py", line 301, in main
    rastr_plot.plot_raster_grid(figname=figname)
  File "D:/NutSync/code/funcs_lib_dg/plot_func/plot_raster_pygmt.py", line 99, in plot_raster_grid
    cmap=cpt
  File "C:\ProgramData\Anaconda3\lib\site-packages\pygmt\helpers\decorators.py", line 586, in new_module
    return module_func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pygmt\helpers\decorators.py", line 726, in new_module
    return module_func(*args, **kwargs)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pygmt\src\grdimage.py", line 173, in grdimage
    fname = stack.enter_context(file_context)
  File "C:\ProgramData\Anaconda3\lib\contextlib.py", line 427, in enter_context
    result = _cm_type.__enter__(cm)
  File "C:\ProgramData\Anaconda3\lib\contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pygmt\clib\session.py", line 1341, in virtualfile_from_grid
    self.put_matrix(gmt_grid, matrix)
  File "C:\ProgramData\Anaconda3\lib\site-packages\pygmt\clib\session.py", line 895, in put_matrix
    raise GMTCLibError(f"Failed to put matrix of type {matrix.dtype}.")
pygmt.exceptions.GMTCLibError: Failed to put matrix of type uint8.
```

**System information**

 I use pygmt 0.5.0 with python 3.7.

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

Support numpy data types `np.int8`, `np.int16`, `np.uint8`, and `np.uint16`.

**References:** 
- https://numpy.org/doc/stable/reference/arrays.scalars.html?highlight=uint32#sized-aliases
- https://github.com/GenericMappingTools/gmt/blob/b552be3a3f76455cd45f32202d468d14ce93cd28/src/gmt_resources.h#L83

Fixes #1960.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by weiji14** [comment]
Awesome @seisman! Didn't realize GMT even supported these other dtypes!

Just need to update some tests, e.g. these lines:

https://github.com/GenericMappingTools/pygmt/blob/1067fa316b214d470f26f4affba743a4835a53c8/pygmt/tests/test_clib_put_matrix.py#L18

https://github.com/GenericMappingTools/pygmt/blob/1067fa316b214d470f26f4affba743a4835a53c8/pygmt/tests/test_clib_put_matrix.py#L64

https://github.com/GenericMappingTools/pygmt/blob/1067fa316b214d470f26f4affba743a4835a53c8/pygmt/tests/test_clib_put_vector.py#L19

https://github.com/GenericMappingTools/pygmt/blob/1067fa316b214d470f26f4affba743a4835a53c8/pygmt/tests/test_clib_put_vector.py#L59

Also 6 more lines in test_clib.py. Maybe a good idea to make the `dtypes` list into a pytest fixture, use something like:

Use `dtypes = "int8 int16 int32 int64 uint8 uint16 uint32 uint64 float32 float64".split()`

File: pygmt/clib/session.py [hunk 1]
```diff
@@ -856,8 +860,8 @@ def put_matrix(self, dataset, matrix, pad=0):
         The dataset must be created by :meth:`pygmt.clib.Session.create_data`
         first. Use ``|GMT_VIA_MATRIX'`` in the family.
 
-        Not at all numpy dtypes are supported, only: float64, float32, int64,
-        int32, uint64, and uint32.
+        Not at all numpy dtypes are supported, only: int8, int16, int32, int64,
+        uint8, uint16, uint32, uint64, float32, float64, str\_ and datetime64.
```
Comment (seisman) [inline pygmt/clib/session.py:864]: I just realized that passing `str_` type matrix to GMT is allowed in the PyGMT `put_matrix` function, but was not documented here. But I'm wondering what happens if passing a `str` matrix to GMT.

Comment (seisman) [inline pygmt/clib/session.py:864]:
```suggestion
        Not at all numpy dtypes are supported, only: int8, int16, int32, int64,
        uint8, uint16, uint32, uint64, float32 and float64.
```

**Comment by seisman** [comment]
Looking at the source code of the `GMT_Put_Matrix` function (https://github.com/GenericMappingTools/gmt/blob/b552be3a3f76455cd45f32202d468d14ce93cd28/src/gmt_api.c#L15458) and the `gmtapi_insert_vector ` function (https://github.com/GenericMappingTools/gmt/blob/b552be3a3f76455cd45f32202d468d14ce93cd28/src/gmt_api.c#L15297), I think only numeric data types are supported.

**Review by weiji14 — APPROVED** [review]
Looks ok to me, but maybe wait and see if @De-Guo has time to verify that it fixes his issue at #1960.

**Comment by seisman** [comment]
> Just as an aside, is there any chance of GMT supporting half precision floats? I.e. [numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float16](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.float16), or 2-byte floats. Not sure if it's possible in C, but supporting float16 will be useful for some machine learning workloads that use float16.

I have no idea about float16. Better to open a feature request in the GMT repository.

**Pull request merged**